File: DetectLoopInLinkedList/DetectLoopInLinkedList.cpp
// DetectLoopInLinkedList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

//#define __STL_SET_ALLOWED__
#ifndef __STL_SET_ALLOWED__
#include <vector>
#else
#include <unordered_set>
#endif

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

void PrintList(ListNode* pHead)
{
    std::cout << pHead->data;
    pHead = pHead->next;

    while (pHead != nullptr)
    {
        std::cout << ", " << pHead->data;
        pHead = pHead->next;
    }
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

ListNode* GetNodeByIndex(ListNode* pHead, int index)
{
    if (index == -1)
        return nullptr;

    for (int i = 0; i < index; ++i)
    {
        pHead = pHead->next;
    }

    return pHead;
}

#ifndef __STL_SET_ALLOWED__
bool NodeExists(std::vector<void*>& nodes, ListNode* node)
{
    for (auto n : nodes)
    {
        if (n == node)
            return true;
    }
    return false;
}

ListNode* StartingNodeOfLoop(ListNode* list)
{
    std::vector<void*> nodes;
    while (list != nullptr)
    {
        if (NodeExists(nodes, list))
            return list;

        nodes.push_back(list);
        list = list->next;
    }

    return nullptr;
}
#else
ListNode* StartingNodeOfLoop(ListNode* list)
{
    std::unordered_set<ListNode*> nodes;

    while (list != nullptr)
    {
        if (nodes.contains(list))
            return list;

        nodes.insert(list);
        list = list->next;
    }

    return nullptr;
}
#endif

ListNode* GetStartingNodeOfLoop(ListNode* list)
{
    return StartingNodeOfLoop(list);
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of nodes in list: ";
        std::cin >> count;
        if (count == 0)
            break;

        ListNode* pHead = nullptr;
        ListNode* pTail = GetListFromInput(&pHead, count);

        std::cout << "The index of starting node of the loop: ";
        int index = -1;
        std::cin >> index;

        pTail->next = GetNodeByIndex(pHead, index);

        ListNode* pNode = GetStartingNodeOfLoop(pHead);
        std::cout << "Loop starts at node: " << (pNode != nullptr ? pNode->data : -1) << std::endl;
        pTail->next = nullptr;
        FreeList(pHead);
    }
}

File: Go/DetectLoopInLinkedList.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) (*ListNode, *ListNode) {
	if len(v) == 0 {
		return nil, nil
	}

	if len(v) == 1 {
		pNode := &ListNode{v[0], nil}
		return pNode, pNode
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead, pCur
}

func GetNodeByIndex(pHead *ListNode, index int) *ListNode {
	if index == -1 {
		return nil
	}

	for i := 0; i < index; i++ {
		if pHead == nil {
			return pHead
		}
		pHead = pHead.next
	}

	return pHead
}

func GetStartingNodeOfLoop(list *ListNode) *ListNode {
	checked := make(map[*ListNode]struct{})
	for ; list != nil; list = list.next {
		_, exists := checked[list]
		if exists {
			return list
		}
		checked[list] = struct{}{}
	}

	return nil
}

func main() {
	for {
		fmt.Print("Number of elements: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := GetArrayFromInput(count, "Please enter the list: ")
		pHead, pTail := GetListFromArray(vec)

		fmt.Print("The node index the loop start with: ")
		var m int
		fmt.Scan(&m)

		pTail.next = GetNodeByIndex(pHead, m)

		pNode := GetStartingNodeOfLoop(pHead)
		if pNode != nil {
			fmt.Printf("The loop starts from %d\n", pNode.data)
		} else {
			fmt.Println("There is not a loop in the list")
		}
	}
}
